File: app/src/main/java/com/example/megaport/mynews/Controllers/Utils/Notifications/AlarmReceiver.java
package com.example.megaport.mynews.Controllers.Utils.Notifications;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.support.v4.app.NotificationCompat;
import android.util.Log;

import com.example.megaport.mynews.Controllers.Activities.SearchResultActivity;
import com.example.megaport.mynews.Controllers.Utils.MyNewsStreams;
import com.example.megaport.mynews.Models.notifications.MyNewsResultAPI;
import com.example.megaport.mynews.Models.notifications.NotificationsPreferences;
import com.example.megaport.mynews.R;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import static android.content.Context.MODE_PRIVATE;
import static com.example.megaport.mynews.Controllers.Activities.NotificationsActivity.NOTIFICATIONS_STATE;
import static com.example.megaport.mynews.Controllers.Activities.NotificationsActivity.PREFS;

public class AlarmReceiver extends BroadcastReceiver {

    private static final String NOTIFICATION_TITLE = "MyNews";
    private static final String NOTIFICATION_MESSAGE = "Articles that may interest you!";

    private Context mContext;

    private static final String QUERY = "QUERY";
    private static final String FILTER_QUERY = "FILTER_QUERY";
    private static final String BEGIN_DATE = "BEGIN_DATE";

    @Override
    public void onReceive(Context context, Intent intent) {
        this.mContext = context;
        this.retrieveSharedPreferences();
    }

    private void retrieveSharedPreferences(){
        SharedPreferences preferences = mContext.getSharedPreferences(PREFS, MODE_PRIVATE);
        Gson gson = new Gson();
        Type type = new TypeToken<NotificationsPreferences>(){}.getType();
        String jsonState = preferences.getString(NOTIFICATIONS_STATE,"");
        NotificationsPreferences notificationsPreferences = gson.fromJson(jsonState, type);
        if (notificationsPreferences != null){
            String queryTerm = notificationsPreferences.getQueryTerm();
            List<String> categoryList = notificationsPreferences.getCategoryList();
            this.executeHttpRequestWithRetrofit(queryTerm, categoryList);
        }
    }

    private void executeHttpRequestWithRetrofit(String queryTerm, List<String> categoryList){
        Disposable disposable = MyNewsStreams.streamFetchNotifications( queryTerm, categoryList, todayDate(), null ).subscribeWith(new DisposableObserver<MyNewsResultAPI>() {
            @Override
            public void onNext(MyNewsResultAPI nyTimesResultAPI) {
                if (!(MyNewsResultAPI.getResponse().getDocs().isEmpty())) {
                    sendNotifications( queryTerm, categoryList );
                }
            }

            @Override
            public void onError(Throwable e) {
                Log.e("TAG", "Error request");
            }

            @Override
            public void onComplete() {
            }
        } );
    }

    private void sendNotifications(String queryTerm, List<String> categoryList){
        //Intent to invoke app when click on notification.
        Intent intentToRepeat = new Intent(mContext, SearchResultActivity.class);
        intentToRepeat.putExtra(QUERY,queryTerm);
        intentToRepeat.putExtra(BEGIN_DATE,todayDate());
        intentToRepeat.putStringArrayListExtra(FILTER_QUERY,(ArrayList<String>)categoryList);
        //set flag to restart/relaunch the app
        intentToRepeat.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);

        //Pending intent to handle launch of Activity in intent above
        PendingIntent pendingIntent = PendingIntent.getActivity(mContext, NotificationHelper.ALARM_TYPE_RTC, intentToRepeat, PendingIntent.FLAG_UPDATE_CURRENT);

        //Build notification
        Notification repeatedNotification = buildLocalNotification(mContext, pendingIntent).build();

        //Send local notification
        NotificationHelper.getNotificationManager(mContext).notify(NotificationHelper.ALARM_TYPE_RTC, repeatedNotification);
    }

    private NotificationCompat.Builder buildLocalNotification(Context context, PendingIntent pendingIntent) {

        return new NotificationCompat.Builder(context)
                .setContentIntent(pendingIntent)
                .setSmallIcon(R.drawable.ic_openclassrooms)
                .setContentTitle(NOTIFICATION_TITLE)
                .setContentText(NOTIFICATION_MESSAGE)
                .setAutoCancel(true);
    }

    private String todayDate(){
        @SuppressLint("SimpleDateFormat") DateFormat df = new SimpleDateFormat("yyyyMMdd");
        return df.format(Calendar.getInstance().getTime());
    }
}

File: app/src/main/java/com/example/megaport/mynews/controllers/utils/notifications/AlarmReceiver.kt
package com.example.megaport.mynews.controllers.utils.notifications

import android.annotation.SuppressLint
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.core.app.NotificationCompat
import android.util.Log
import com.example.megaport.mynews.controllers.utils.MyNewsStreams
import com.example.megaport.mynews.models.notifications.MyNewsResultAPI
import com.example.megaport.mynews.models.notifications.NotificationsPreferences
import com.example.megaport.mynews.R
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.text.SimpleDateFormat
import io.reactivex.observers.DisposableObserver
import android.content.Context.MODE_PRIVATE
import android.graphics.Color
import com.example.megaport.mynews.controllers.activities.NotificationsActivity.Companion.NOTIFICATIONS_STATE
import com.example.megaport.mynews.controllers.activities.NotificationsActivity.Companion.PREFS
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.addTo
import java.util.*

class AlarmReceiver : BroadcastReceiver() {

    private var mContext: Context? = null
    private var numberOfArticle: Int? = 0

    private val disposable: CompositeDisposable by lazy {
        CompositeDisposable()
    }

    @SuppressLint("UnsafeProtectedBroadcastReceiver")
    override fun onReceive(context: Context, intent: Intent) {
        this.mContext = context
        this.retrieveSharedPreferences()
    }

    private fun retrieveSharedPreferences() {
        val preferences = mContext!!.getSharedPreferences(PREFS, MODE_PRIVATE)
        val gson = Gson()
        val type = object : TypeToken<NotificationsPreferences>() {

        }.type
        val jsonState = preferences.getString(NOTIFICATIONS_STATE, "")
        val notificationsPreferences = gson.fromJson<NotificationsPreferences>(jsonState, type)
        if (notificationsPreferences != null) {
            val queryTerm = notificationsPreferences.queryTerm
            val categoryList = notificationsPreferences.categoryList
            this.executeHttpRequestWithRetrofit(queryTerm, categoryList)
        }
    }

    private fun executeHttpRequestWithRetrofit(queryTerm: String?, categoryList: List<String>?) {
        MyNewsStreams.streamFetchNotifications(queryTerm, categoryList, todayDate(), null).subscribeWith(object : DisposableObserver<MyNewsResultAPI>() {
            override fun onNext(nyTimesResultAPI: MyNewsResultAPI) {
                if (nyTimesResultAPI.response?.docs?.isEmpty() == false) {
                    numberOfArticle = nyTimesResultAPI.response.docs.size
                    sendNotifications()
                }
            }

            override fun onError(e: Throwable) {
                Log.e("TAG", "Error request")
            }

            override fun onComplete() {}
        }).addTo(disposable)
    }

    private fun sendNotifications() {

        //Build notification
        val repeatedNotification = mContext?.let { buildLocalNotification(it).build() }

        val notificationManager = mContext!!.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val notificationChannel = NotificationChannel(NOTIFICATION_CHANNEL_ID, NOTIFICATION_CHANNEL_NAME, importance)
            notificationChannel.enableLights(true)
            notificationChannel.lightColor = Color.RED
            notificationChannel.enableVibration(true)
            buildLocalNotification(mContext!!).setChannelId(NOTIFICATION_CHANNEL_ID)
            Objects.requireNonNull<NotificationManager>(notificationManager).createNotificationChannel(notificationChannel)
        }
        Objects.requireNonNull<NotificationManager>(notificationManager).notify(NotificationHelper.ALARM_TYPE_RTC, repeatedNotification)
    }

    private fun buildLocalNotification(context: Context): NotificationCompat.Builder {

        return NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_openclassrooms)
                .setContentTitle(NOTIFICATION_TITLE)
                .setContentText("$numberOfArticle articles may interest you!")
                .setAutoCancel(true)
    }

    private fun todayDate(): String {
        @SuppressLint("SimpleDateFormat") val df = SimpleDateFormat("yyyyMMdd")
        return df.format(Calendar.getInstance().time)
    }

    companion object {

        private const val NOTIFICATION_TITLE = "MyNews"
        private val NOTIFICATION_CHANNEL_ID = "5000"
        private val NOTIFICATION_CHANNEL_NAME = "MyNews"


    }
}